For PostgreSQL >= 9.4 use `gen_random_uuid()`

yawboakye opened the pull request:
Since version 9.4, the PostgreSQL community recommends moving away from
`uuid-ossp`'s UUID generation functions and using `pcrypto`'s
`gen_random_uuid()` instead. It generates UUID version 4 by default,
which is what everyone tends to use by the way.

These changes use the appropriate UUID function depending on the
underlying PostgreSQL server's version, while maintaining
`uuid_generate_v4()` in older migrations.

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_definitions.rb:
```diff
@@ -11,27 +11,31 @@ module ColumnMethods
         def primary_key(name, type = :primary_key, **options)
-          options[:default] = options.fetch(:default, 'uuid_generate_v4()') if type == :uuid
+          if type == :uuid
+            uuid_function = ActiveRecord::Base.connection.supports_pgcrypto_uuid? ? 'gen_random_uuid()' : 'uuid_generate_v4()'
```
This should be using the connection that it already hold instead of `ActiveRecord::Base`. Could you change it?

yawboakye replied:
@rafaelfranca Help me out. I can't find which `connection` to use :(

rafaelfranca replied:
Seems `TableDefinition` doesn't have access to a connection. So we should remove the connection check here and pick `gen_random_uuid` as default. Using `ActiveRecord::Base.connection` will make this code not support multi connections.

yawboakye replied:
How will that work for users on PostgreSQL 9.3 or lower since `gen_random_uuid()` is not available to them? Will `9.4` be the minimum required version?

rafaelfranca replied:
They will have to explicitly tell that they want to use `uuid_generate_v4()`. I think it is a fair trade off and since it is documented in the changelog users will know it.

rafaelfranca commented on activerecord/test/cases/adapters/postgresql/uuid_test.rb:
```diff
@@ -195,17 +210,52 @@ def test_pk_and_sequence_for_uuid_primary_key
+    else
+      def test_schema_dumper_for_uuid_primary_key
+        schema = dump_table_schema "pg_uuids"
+        assert_match(/\bcreate_table "pg_uuids", id: :uuid, default: -> { "uuid_generate_v4\(\)" }/, schema)
```
why is this `uuid_generate_v4` not `uuid_generate_v1`

yawboakye replied:
Two different columns. The `id` column, also the primary key, uses the `uuid_generate_v1()` function while `other_uuid` uses `uuid_generate_v4()`. Both are functions in the `uuid-ossp` extension.

rafaelfranca replied:
Yeah, but in this test `if` is using `uuid_generate_v4` while in the test above (that should be the same test of this) is using `uuid_generate_v1`

yawboakye replied:
Thanks for pointing it out. Turns out that test isn't necessary. The test for using `gen_random_uuid()` when no `default` is set is done using the `pg_uuids_3` table instead.

kamipo commented:
Duplicate to #21327.

yawboakye commented:
@kamipo Yes. This PR started life as corrections to @y-yagi's original.

ianks commented:
@yawboakye Would it be possible to include some information about migrating from `uuid-ossp` to `gen_random_uuid()`?

yawboakye committed:
For `PostgreSQL >= 9.4` use `gen_random_uuid()`

Since 9.4, PostgreSQL recommends using `pgcrypto`'s `gen_random_uuid()`
to generate version 4 UUIDs instead of the functions in the `uuid-ossp`
extension.

These changes uses the appropriate UUID function depending on the
underlying PostgreSQL server's version, while maintaining
`uuid_generate_v4()` in older migrations.

senny commented:
@ianks I don't think we need to cover migration issues. This is a guide about Active Record not the PostgreSQL database. We should just make sure that we encourage best practices in our examples.

yawboakye commented:
@senny As this will be my first contribution to Rails, I want to see it through. So please provide feedback and comments to help me improve it. Thanks cc/ @rafaelfranca @maclover7

rafaelfranca commented:
r? @senny

senny commented on activerecord/test/cases/adapters/postgresql/uuid_test.rb:
```diff
@@ -9,6 +9,10 @@ def connection
   end
+
+  def uuid_function
+    connection.supports_pgcrypto_uuid? ? 'gen_random_uuid()' : 'uuid_generate_v4()'
```
Wouldn't this mean that upgrading the PG database would case migrations to generate a different result? This might be unexpected at times.

Using a configuration option that's set in an initializer could do the trick. Quite a bit of work for a simple change like this though.

 @matthewd any thoughts how we could make this less intrusive?

matthewd replied:
This is just the test, so it seems fine

matthewd commented on activerecord/lib/active_record/migration/compatibility.rb:
```diff
@@ -28,6 +28,11 @@ def timestamps(**options)
+            if options[:id] == :uuid && !options[:default]
+              options[:default] = "uuid_generate_v4()"
+            end
+          end
```
This needs to be in `V5_0`

kamipo commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   For PostgreSQL >= 9.4 use `pgcrypto`'s `gen_random_uuid()` instead of
+    `uuid-ossp`'s UUID generation function.
+
+    *Yaw Boakye*
```
The original of this PR is #21327. I think that it should be credited the original author.

```
    *Yuji Yaginuma*, *Yaw Boakye*
```

yawboakye replied:
@kamipo I disagree. Here's why.

I don't know whether I'll call https://github.com/rails/rails/pull/21327 original. I discovered it when I wanted to submit the improvements above. Some things were wrong and a couple of incorrect statements. So instead of sending another PR I rather left comments for the author's perusal. (See my comments.) I wasn't going to take it on until @senny extended the opportunity to me. And then what started as minor changes turned to a big overhaul, and a constant rebase to keep it relevant. If I still can't credit myself for all the work I've put into it then that's a problem.

yawboakye commented:
@matthewd Hi Matt, looks like we're good to go! 😄

matthewd commented:
🎉

❤️❤️

prathamesh-sonpatki commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_definitions.rb:
```diff
@@ -11,27 +11,28 @@ module ColumnMethods
         # +SecureRandom.uuid+ method and a +before_save+ callback, for instance.
         def primary_key(name, type = :primary_key, **options)
-          options[:default] = options.fetch(:default, "uuid_generate_v4()") if type == :uuid
+          options[:default] = options.fetch(:default, "gen_random_uuid()") if type == :uuid
```
nevermind, I got confused by PR description but checked comment by Rafael.

jcraigk commented on activerecord/test/cases/adapters/postgresql/uuid_test.rb:
```diff
@@ -206,6 +224,33 @@ def test_schema_dumper_for_uuid_primary_key_with_custom_default
+      ensure
+        drop_table "pg_uuids_4"
+      end
+    else
```
Unnecessary `else`?

kamipo replied:
Removed at #27170.

Lewiscowles1986 commented:
Is there a way to override this behaviour? It's a problem

- local dev we have to enable the extension if we use postgres > 9.4
- deploying to RDS we don't want our application user to be able to enable and disable extensions or know about root credentials

I'm trying (perhaps naively) to force using the `uuid_generate_v4()` which seems to be available for all databases we use
